examples/generative/pixelcnn.py

```diff
@@ -0,0 +1,168 @@
+"""
+Title: PixelCNN
+Author: [ADMoreau](https://github.com/ADMoreau)
+Date created: 2020/05/17
+Last modified: 2020/05/23
+Description: PixelCNN implemented in Keras.
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras, nn
+from tensorflow.keras import layers
+
+"""
+## Getting the Data
+"""
+
+# Model / data parameters
+num_classes = 10
+input_shape = (28, 28, 1)
+n_residual_blocks = 5
+# the data, split between train and test sets
+(x, _), (y, _) = keras.datasets.mnist.load_data()
+# Concatenate all of the images together
+data = np.concatenate((x, y), axis=0)
+# round all pixel values less than 33% of the max 256 value to 0
+# anything above this value gets rounded up to 1 so that all values are either
+# 0 or 1
+data = np.where(data < (0.33 * 256), 0, 1)
+data = data.astype(np.float32)
+
+"""
+## Create two Layer classes required by the model
+"""
+
+# The first layer is the PixelCNN layer. This layer simply
+# builds on the 2D convolutional layer, but includes masking.
+class PixelConvLayer(layers.Layer):
+    def __init__(self, mask_type, **kwargs):
+        super(PixelConvLayer, self).__init__()
+        self.mask_type = mask_type
+        self.conv = layers.Conv2D(**kwargs)
+
+    def build(self, input_shape):
+        # Build the conv2d layer to initialize kernel variables
+        self.conv.build(input_shape)
+        # Use the initialized kernel to create the mask
+        kernel_shape = self.conv.kernel.get_shape()
+        self.mask = np.zeros(shape=kernel_shape)
+        self.mask[: kernel_shape[0] // 2, ...] = 1.0
+        self.mask[kernel_shape[0] // 2, : kernel_shape[1] // 2, ...] = 1.0
+        if self.mask_type == "B":
+            self.mask[kernel_shape[0] // 2, kernel_shape[1] // 2, ...] = 1.0
+
+    def call(self, inputs):
+        self.conv.kernel.assign(self.conv.kernel * self.mask)
+        return self.conv(inputs)
+
+
+# Next, we build our residual block layer.
+# This is just a normal residual block, but base don the PixelConvLayer.
+class ResidualBlock(keras.layers.Layer):
+    def __init__(self, filters, **kwargs):
+        super(ResidualBlock, self).__init__(**kwargs)
+        self.activation = keras.layers.ReLU()
```

Comment:
As far as I can tell you don't need the block to start with a ReLU, because 1) it ends with a ReLU, 2) you already apply a ReLU before the first block instance when you create the model below. So this first ReLU seems redundant